application/src/main/java/org/opentripplanner/street/model/edge/StreetEdge.java

```diff
@@ -987,35 +1031,39 @@ static int defaultMillimeterLength(LineString geometry) {
    * zone applies to where we pick up a vehicle with a specific network.
    */
   private State[] splitStatesAfterHavingExitedNoDropOffZoneWhenReverseSearching(State s0) {
-    var networks = Stream.concat(
+    var preferences = s0.getRequest().preferences();
+    var states = tov
+      .rentalRestrictions()
+      .noDropOffNetworks()
+      .stream()
+      .map(network ->
+        makeStateAfterHavingExitedNoDropOffZoneWhenReverseSearching(s0, network, preferences)
+      )
+      .filter(Objects::nonNull)
+      .toList();
+    var statesStream = states.stream();
+
+    if (!states.isEmpty()) {
       // null is a special rental network that speculatively assumes that you can take any vehicle
       // you have to check in the rental edge if this has search has been started in a no-drop off zone
-      Stream.of((String) null),
-      tov.rentalRestrictions().noDropOffNetworks().stream()
-    );
-
-    var states = networks.map(network -> {
-      var edit = doTraverse(s0, TraverseMode.WALK, false);
-      if (edit != null) {
-        edit.dropFloatingVehicle(s0.vehicleRentalFormFactor(), network, s0.getRequest().arriveBy());
-        if (network != null) {
-          edit.resetStartedInNoDropOffZone();
-        }
-        return edit.makeState();
-      }
-      return null;
-    });
+      statesStream =
+        Stream.concat(
+          Stream.of(
+            makeStateAfterHavingExitedNoDropOffZoneWhenReverseSearching(s0, null, preferences)
+          ),
+          statesStream
+        );
+    }
 
     // Also include a state which continues walking, because the vehicle rental states are
     // speculation. It is possible that the rental states don't end up at the target at all
     // due to mode limitations or not finding a place to pick up the rental vehicle, or that
     // the rental possibility is simply more expensive than walking.
     StateEditor walking = doTraverse(s0, TraverseMode.WALK, false);
     if (walking != null) {
-      var forkState = walking.makeState();
-      states = Stream.concat(Stream.of(forkState), states);
+      statesStream = Stream.concat(Stream.of(walking.makeState()), statesStream);
     }
-    return State.ofStream(states);
+    return State.ofStream(statesStream);
```

Comment:
Btw I wonder if using streams in edge traversal is inefficient, although I guess this method is rarely called.

Reply:
The overhead of using streams over for-loops is around 10 nanoseconds per iteration per method call in the stream chain: https://www.baeldung.com/java-streams-vs-loops

In my opinion, it is a readability issue first and foremost, although here streams were used before me so I just continued, but I attempted a for-loop and it was a mess.

Reply:
Instead of the looping performance, I'm more concerned about the initialization cost of streams and conversion to array.  Seems like we are also doing multiple filtering of null objects away from the stream but that probably at least gets optimized away automatically. I don't think we should try to refactor this unless we have concrete evidence that we are spending a lot of time on streams.

Reply:
Ok, I tried again and now the for loop is readable.